Fix empty network config

eiabea opened the pull request:
**What this PR does / why we need it**:
Provisioning a VM without defining a `Network` in the `providerConfig` object results in a nil pointer exception

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->
/kind bug

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Anexia Provider: Fix nil pointer exception when providing no network configuration
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

eiabea committed:
Check if providerConfig is set

Add test for empty network provider config

Signed-off-by: Manuel Zangl <mzangl@anexia-it.com>

marioreggiori reviewed (commented):
Looking mostly good.
Do you think we could remove some duplication in the test file similar to how "Test resolve template" uses test cases?

eiabea committed:
Refactor provision vm test cases

Create test cases to reduce code duplication
Move common types and functions to helper_test

Signed-off-by: Manuel Zangl <mzangl@anexia-it.com>

marioreggiori commented on pkg/cloudprovider/provider/anexia/provider_test.go:
```diff
@@ -66,6 +66,19 @@ func TestAnexiaProvider(t *testing.T) {
 	a.FakeExisting(&vspherev1.Template{Identifier: "WRONG-TEMPLATE-NAME", Name: "Wrong Template Name", Build: "b02"})
+	a.FakeExisting(&vspherev1.Template{Identifier: "TEMPLATE-ID-NO-NETWORK-CONFIG", Name: "no-network-config", Build: "b03"})
+
+	testhelper.Mux.HandleFunc("/api/ipam/v1/address/reserve/ip/count.json", func(writer http.ResponseWriter, request *http.Request) {
+		err := json.NewEncoder(writer).Encode(address.ReserveRandomSummary{
+			Data: []address.ReservedIP{
+				{
+					ID:      "IP-ID",
+					Address: "8.8.8.8",
+				},
+			},
+		})
+		testhelper.AssertNoErr(t, err)
+	})
```
Can we move this back into "Test provision VM"?

marioreggiori commented on pkg/cloudprovider/provider/anexia/provider_test.go:
```diff
@@ -74,120 +87,97 @@ func TestAnexiaProvider(t *testing.T) {
+					rc.ProviderConfig = &providerconfigtypes.Config{}
+				}),
+				AssertJsonBody: func(jsonBody jsonObject) {
+					testhelper.AssertEquals(t, jsonBody["dns1"], nil)
+					testhelper.AssertEquals(t, jsonBody["dns2"], nil)
+					testhelper.AssertEquals(t, jsonBody["dns3"], nil)
+					testhelper.AssertEquals(t, jsonBody["dns4"], nil)
```
When serializing this to json there is a difference between `nil` and _not set_. I think we should make sure that this is not set in case no DNS config has been provided (also `dns2` in the 0th test case).

eiabea replied:
I ended up with something like this:

```
_, exists := jsonBody["dns1"]
testhelper.AssertEquals(t, exists, false)
_, exists = jsonBody["dns2"]
testhelper.AssertEquals(t, exists, false)
_, exists = jsonBody["dns3"]
testhelper.AssertEquals(t, exists, false)
_, exists = jsonBody["dns4"]
testhelper.AssertEquals(t, exists, false)
```

Do you have a nicer looking solution?

marioreggiori replied:
Unfortunately not without introducing another test utility package or doing some refactoring (we could use `AssertJSONEquals` instead, but we'd need to check everything then). I think your solution is suitable in this case.

marioreggiori commented on pkg/cloudprovider/provider/anexia/provider_test.go:
```diff
@@ -74,120 +87,97 @@ func TestAnexiaProvider(t *testing.T) {
+
+		for _, testCase := range testCases {
+
+			TemplateID := testCase.ReconcileContext.Config.TemplateID
+			LocationID := testCase.ReconcileContext.Config.LocationID
```
Couldn't get the linter to work. But these should probably be lowercase.

marioreggiori reviewed (changes requested):
Almost there :+1: Good job!

marioreggiori reviewed (approved):
LGTM :+1: 
Remember to sign (& maybe squash) your commits before crating the upstream PR.

LittleFox94 commented:
upstream PR: https://github.com/kubermatic/machine-controller/pull/1646
